### PR TITLE
fix(python): pandas 2.0 compat

### PR DESCRIPTION
**Pull request by MarcoGorelli**

### CODE REVIEW

File: py-polars/tests/unit/test_interop.py [hunk 1]
```diff
@@ -605,10 +605,10 @@ def test_from_empty_arrow() -> None:
 
     # 2705
     df1 = pd.DataFrame(columns=["b"], dtype=float)
-    tbl = pa.Table.from_pandas(df1)
+    tbl = pa.Table.from_pandas(df1, preserve_index=True)
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_interop.py:608]: empty dataframes now have a rangeindex by default: https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#empty-dataframes-series-will-now-default-to-have-a-rangeindex

in pyarrow https://arrow.apache.org/docs/python/generated/pyarrow.Table.html :

> The default of None will store the index as a column, except for RangeIndex which is stored as metadata only. Use preserve_index=True to force it to be stored as a column.

File: py-polars/polars/utils/_construction.py [hunk 2]
```diff
@@ -1327,7 +1327,7 @@ def to_frame_chunk(values: list[Any], schema: SchemaDefinition | None) -> DataFr
 
 def pandas_has_default_index(df: pd.DataFrame) -> bool:
     """Identify if the pandas frame only has a default (or equivalent) index."""
-    from pandas.core.indexes.numeric import IntegerIndex
+    from pandas import Index
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/_construction.py:1330]: https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#index-can-now-hold-numpy-numeric-dtypes

File: py-polars/tests/unit/test_interop.py [hunk 3]
```diff
@@ -491,7 +491,7 @@ def test_cat_to_pandas() -> None:
 
     try:
         pd_pa_out = df.to_pandas(use_pyarrow_extension_array=True)
-        assert pd_pa_out["a"].dtype.type == pa.DictionaryType
+        assert pd_pa_out["a"].dtype.type == pd.core.dtypes.dtypes.CategoricalDtypeType
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_interop.py:494]: I don't know why this changed, looking into it

File: py-polars/polars/utils/_construction.py [hunk 4]
```diff
@@ -1341,10 +1340,7 @@ def pandas_has_default_index(df: pd.DataFrame) -> bool:
     else:
         # finally, is the index _equivalent_ to a default unnamed
         # integer index with frame data that was previously sorted
-        return (
-            isinstance(df.index, IntegerIndex)
-            and (df.index.sort_values() == np.arange(len(df))).all()
-        )
+        return df.index.sort_values() == np.arange(len(df)).all()
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/_construction.py:1343]: IntegerIndex is no more https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#index-can-now-hold-numpy-numeric-dtypes

File: py-polars/tests/unit/test_interop.py [hunk 5]
```diff
@@ -604,7 +604,7 @@ def test_from_empty_arrow() -> None:
     assert df.dtypes == [pl.Float64, pl.Float64]
 
     # 2705
-    df1 = pd.DataFrame(columns=["b"], dtype=float)
+    df1 = pd.DataFrame(columns=["b"], dtype=float, index=pd.Index([]))
     tbl = pa.Table.from_pandas(df1)
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_interop.py:608]: empty dataframes now have a RangeIndex by default: https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#empty-dataframes-series-will-now-default-to-have-a-rangeindex

in pyarrow https://arrow.apache.org/docs/python/generated/pyarrow.Table.html :

> The default of None will store the index as a column, except for RangeIndex which is stored as metadata only. Use preserve_index=True to force it to be stored as a column.

**Comment by ritchie46** [comment]
Thanks for the quick fix! 💪

**Comment by jakob-keller** [comment]
FWIW: After this change, one of the unit tests kept failing, until I manually updated pandas in my local environment. In hindsight that was to be expected, but maybe someone else runs into this as well.

Could the `Makefile` be tuned to automatically install any dependency updates? Maybe the pandas entry in `requirements-dev.txt` needs to be restricted to >=2 from now on?

**Pull request merged**